internal/datastore/common/gc.go

```diff
@@ -149,14 +149,20 @@ var MaxGCInterval = 60 * time.Minute
 // StartGarbageCollector loops forever until the context is canceled and
 // performs garbage collection on the provided interval.
 func StartGarbageCollector(ctx context.Context, collectable GarbageCollectableDatastore, interval, window, timeout time.Duration) error {
-	return startGarbageCollectorWithMaxElapsedTime(ctx, collectable, interval, window, 0, timeout, gcFailureCounter)
+	return runPeriodicallyWithBackoff(ctx, func() error {
+		// NOTE: we're okay using the parent context here because the
+		// callers of this function create a dedicated garbage collection
+		// context anyway, which is only cancelled when the ds is closed.
+		gcCtx, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
+		return RunGarbageCollection(gcCtx, collectable, window)
+	}, interval, window, timeout, gcFailureCounter)
 }
 
-func startGarbageCollectorWithMaxElapsedTime(ctx context.Context, collectable GarbageCollectableDatastore, interval, window, maxElapsedTime, timeout time.Duration, failureCounter prometheus.Counter) error {
+func runPeriodicallyWithBackoff(ctx context.Context, taskFn func() error, interval, window, timeout time.Duration, failureCounter prometheus.Counter) error {
```

Comment:
This is another big change - the `runPeriodicallyWithBackoff` logic is mostly used by GC, but there's nothing GC-specific about it, so we can test it separately and then let them be composed.